tests/ga/test_policy_engine.py

```diff
@@ -15,20 +15,579 @@
+        Cases should be a list of policies.
+        For each policy in the list, we create a policy file, initialize policy engine, and assert that InvalidPolicyError
+        is raised.
+        """
+        for policy in cases:
+            self._create_policy_file(policy)
+            msg = "invalid policy should not have parsed successfully: {0}. Policy: {1}".format(assert_msg, policy)
+            with self.assertRaises(InvalidPolicyError, msg=msg):
+                _PolicyEngine()
+
+class TestPolicyEngine(_TestPolicyBase):
+    """
+    Test policy enablement and parsing logic for _PolicyEngine.
+    """
+    def test_policy_enforcement_should_be_enabled_when_policy_file_exists_and_conf_flag_true(self):
+        """
+        When conf flag is set to true and policy file is present at expected location, feature should be enabled.
+        """
+        # Create policy file with empty policy object at the expected path to enable feature.
+        self._create_policy_file({})
+        engine = _PolicyEngine()
+        self.assertTrue(engine.policy_enforcement_enabled,
+                        msg="Conf flag is set to true so policy enforcement should be enabled.")
+
+    def test_policy_enforcement_should_be_disabled_when_conf_flag_false_or_no_policy_file(self):
 
+        # Test when conf flag is turned off - feature should be disabled.
+        self.patch_conf_flag.stop()
+        engine1 = _PolicyEngine()
+        self.assertFalse(engine1.policy_enforcement_enabled,
+                         msg="Conf flag is set to false and policy file missing so policy enforcement should be disabled.")
 
-class TestPolicyEngine(AgentTestCase):
-    def test_policy_enforcement_should_be_enabled(self):
-        with patch('azurelinuxagent.ga.policy.policy_engine.conf.get_extension_policy_enabled', return_value=True):
-            engine = PolicyEngine()
-            self.assertTrue(engine.is_policy_enforcement_enabled(),
-                            msg="Conf flag is set to true so policy enforcement should be enabled.")
+        # Turn on conf flag - feature should still be disabled, because policy file is not present.
+        self.patch_conf_flag.start()
+        engine2 = _PolicyEngine()
+        self.assertFalse(engine2.policy_enforcement_enabled,
+                         msg="Policy file is not present so policy enforcement should be disabled.")
 
-    def test_policy_enforcement_should_be_disabled(self):
-        engine = PolicyEngine()
-        self.assertFalse(engine.is_policy_enforcement_enabled(),
+        # Create a policy file, but turn off conf flag - feature should be disabled due to flag.
+        self.patch_conf_flag.stop()
+        self._create_policy_file({})
+        engine3 = _PolicyEngine()
+        self.assertFalse(engine3.policy_enforcement_enabled,
                          msg="Conf flag is set to false so policy enforcement should be disabled.")
 
+    def test_policy_should_be_enabled_even_if_policy_file_deleted(self):
```

Comment:
What is the intention of this test? it is correct, but I do not the value it adds. existence of the file (and its contents) are loaded only once, at initialization time. if the file is removed, the enabled flag will still be true since it is set at initialization, but the contents of the policy will still be loaded, too. Yet, we check only the enable flag.

I guess the above is just a long way to sugest "let's remove the test unless we have a specific concern about deleting the file and the value of the enabled flag"

Reply:
Removed